File: backend/src/routes/categories.js
import express from 'express';
import { v4 as uuidv4 } from 'uuid';
import { query, getOne, getMany } from '../db.js';
import { authMiddleware } from '../middleware/auth.js';
import { broadcast } from '../server.js';

const router = express.Router();

// All routes require authentication
router.use(authMiddleware);

// Get all categories
router.get('/', async (req, res) => {
  try {
    const { type } = req.query;

    let sql = 'SELECT * FROM budget_categories WHERE created_by = $1';
    const params = [req.user.id];

    if (type) {
      sql += ' AND type = $2';
      params.push(type);
    }

    sql += ' ORDER BY name';

    const categories = await getMany(sql, params);
    res.json(categories);
  } catch (error) {
    console.error('Get categories error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get single category
router.get('/:id', async (req, res) => {
  try {
    const category = await getOne(
      'SELECT * FROM budget_categories WHERE id = $1 AND created_by = $2',
      [req.params.id, req.user.id]
    );

    if (!category) {
      return res.status(404).json({ error: 'Category not found' });
    }

    res.json(category);
  } catch (error) {
    console.error('Get category error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Create category
router.post('/', async (req, res) => {
  try {
    const { name, type, icon, color, is_active } = req.body;

    if (!name || !type) {
      return res.status(400).json({ error: 'Name and type are required' });
    }

    const id = uuidv4();
    const now = new Date();

    const category = await getOne(
      `INSERT INTO budget_categories (id, name, type, icon, color, is_active, created_by, created_date)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
       RETURNING *`,
      [id, name, type, icon || '💡', color || null, is_active !== false, req.user.id, now]
    );

    // Broadcast update via WebSocket
    broadcast({ type: 'category_created', data: category, userId: req.user.id });

    res.status(201).json(category);
  } catch (error) {
    console.error('Create category error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Update category
router.put('/:id', async (req, res) => {
  try {
    const { name, type, icon, color, is_active } = req.body;

    // Check if category exists and belongs to user
    const existing = await getOne(
      'SELECT * FROM budget_categories WHERE id = $1 AND created_by = $2',
      [req.params.id, req.user.id]
    );

    if (!existing) {
      return res.status(404).json({ error: 'Category not found' });
    }

    const category = await getOne(
      `UPDATE budget_categories
       SET name = $1, type = $2, icon = $3, color = $4, is_active = $5
       WHERE id = $6
       RETURNING *`,
      [
        name ?? existing.name,
        type ?? existing.type,
        icon ?? existing.icon,
        color ?? existing.color,
        is_active !== undefined ? is_active : existing.is_active,
        req.params.id
      ]
    );

    // Broadcast update via WebSocket
    broadcast({ type: 'category_updated', data: category, userId: req.user.id });

    res.json(category);
  } catch (error) {
    console.error('Update category error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Delete category
router.delete('/:id', async (req, res) => {
  try {
    // Check if category exists and belongs to user
    const existing = await getOne(
      'SELECT * FROM budget_categories WHERE id = $1 AND created_by = $2',
      [req.params.id, req.user.id]
    );

    if (!existing) {
      return res.status(404).json({ error: 'Category not found' });
    }

    await query('DELETE FROM budget_categories WHERE id = $1', [req.params.id]);

    // Broadcast update via WebSocket
    broadcast({ type: 'category_deleted', data: { id: req.params.id }, userId: req.user.id });

    res.json({ message: 'Category deleted successfully' });
  } catch (error) {
    console.error('Delete category error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: fastapi_app/routes/categories.py
"""
Category routes.
"""
import uuid
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status, Query
from models import (
    CategoryCreate, CategoryUpdate, CategoryResponse, TransactionType, MessageResponse
)
from auth import get_current_user, User
from database import query, query_one, execute

router = APIRouter(prefix="/categories", tags=["Categories"])


@router.get("", response_model=List[CategoryResponse])
async def list_categories(
    type: Optional[TransactionType] = Query(None, description="Filter by category type"),
    current_user: User = Depends(get_current_user)
):
    """
    List all budget categories for the authenticated user.

    Supports filtering by:
    - **type**: 'income' or 'expense'
    """
    sql = """
        SELECT id, name, type, icon, color, is_active, created_by, created_date, updated_date
        FROM budget_categories
        WHERE created_by = %s
    """
    params = [current_user.id]

    if type:
        sql += " AND type = %s"
        params.append(type.value)

    sql += " ORDER BY name"

    categories = query(sql, tuple(params))
    return [CategoryResponse(**c) for c in categories]


@router.get("/{category_id}", response_model=CategoryResponse)
async def get_category(
    category_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    Get a single category by ID.
    """
    category = query_one(
        """
        SELECT id, name, type, icon, color, is_active, created_by, created_date, updated_date
        FROM budget_categories
        WHERE id = %s AND created_by = %s
        """,
        (category_id, current_user.id)
    )

    if not category:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )

    return CategoryResponse(**category)


@router.post("", response_model=CategoryResponse, status_code=status.HTTP_201_CREATED)
async def create_category(
    category: CategoryCreate,
    current_user: User = Depends(get_current_user)
):
    """
    Create a new budget category.

    - **name**: Category name (e.g., 'Salary', 'Food', 'Transport')
    - **type**: 'income' or 'expense'
    - **icon**: Emoji icon (default: '💡')
    - **color**: Hex color code (e.g., '#4CAF50')
    """
    # Check if category with same name already exists
    existing = query_one(
        "SELECT id FROM budget_categories WHERE name = %s AND created_by = %s",
        (category.name, current_user.id)
    )

    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Category '{category.name}' already exists"
        )

    category_id = str(uuid.uuid4())
    result = execute(
        """
        INSERT INTO budget_categories (id, name, type, icon, color, is_active, created_by)
        VALUES (%s, %s, %s, %s, %s, %s, %s)
        RETURNING *
        """,
        (
            category_id,
            category.name,
            category.type.value,
            category.icon or "💡",
            category.color,
            category.is_active if category.is_active is not None else True,
            current_user.id
        )
    )

    return CategoryResponse(**result)


@router.put("/{category_id}", response_model=CategoryResponse)
async def update_category(
    category_id: str,
    category: CategoryUpdate,
    current_user: User = Depends(get_current_user)
):
    """
    Update an existing category.
    Only provided fields will be updated.
    """
    existing = query_one(
        "SELECT * FROM budget_categories WHERE id = %s AND created_by = %s",
        (category_id, current_user.id)
    )

    if not existing:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )

    # Build update query dynamically
    updates = []
    params = []

    if category.name is not None:
        # Check for duplicate name
        duplicate = query_one(
            "SELECT id FROM budget_categories WHERE name = %s AND created_by = %s AND id != %s",
            (category.name, current_user.id, category_id)
        )
        if duplicate:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Category '{category.name}' already exists"
            )
        updates.append("name = %s")
        params.append(category.name)

    if category.type is not None:
        updates.append("type = %s")
        params.append(category.type.value)

    if category.icon is not None:
        updates.append("icon = %s")
        params.append(category.icon)

    if category.color is not None:
        updates.append("color = %s")
        params.append(category.color)

    if category.is_active is not None:
        updates.append("is_active = %s")
        params.append(category.is_active)

    if not updates:
        return CategoryResponse(**existing)

    updates.append("updated_date = NOW()")
    params.append(category_id)
    params.append(current_user.id)

    sql = f"UPDATE budget_categories SET {', '.join(updates)} WHERE id = %s AND created_by = %s RETURNING *"
    result = execute(sql, tuple(params))

    return CategoryResponse(**result)


@router.delete("/{category_id}", response_model=MessageResponse)
async def delete_category(
    category_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    Delete a category.
    Transactions using this category will have their category_id set to NULL.
    """
    result = execute(
        "DELETE FROM budget_categories WHERE id = %s AND created_by = %s RETURNING id",
        (category_id, current_user.id)
    )

    if not result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )

    return MessageResponse(message="Category deleted successfully")